### PR TITLE
Fix GCC build warning in NetBSD 10

### PR DESCRIPTION
**Pull request by Explorer09**
Thanks to the CI build log provided in here:
https://github.com/htop-dev/htop/actions/runs/8680048533

### CODE REVIEW

**Comment by BenBE** [comment]
This will likely cause mis-classifications of certain characters due to truncating their MSB. Also affects other platforms; looks like a regression to me …

**Comment by Explorer09** [comment]
@BenBE Truncating MSB? What do you mean?

By the way, I referenced these Stack Overflow questions for the fix:
https://stackoverflow.com/questions/9972359
https://stackoverflow.com/questions/18502257

I personally consider this an "artifact" of the C language design when the `char` type can be signed by default by implementation (we are already past the ASCII era when it comes to character encodings). Modern C strings should use `unsigned char` as default for `char` or else we would be unable to process UTF-8 or any other 8-bit encodings.

**Comment by BenBE** [comment]
MSB AKA Most Significant Byte. I just skimmed over it and falsely remembered `data[i]` to by `cchar` AKA full UTF-8. But with `char *data` as the arg this makes sense to have those casts.

Just not sure, how MB (multi-byte) strings like UTF-8 render properly … ;-)

**Comment by Explorer09** [comment]
> MSB AKA Most Significant Byte. I just skimmed over it and falsely remembered `data[i]` to by `cchar` AKA full UTF-8. But with `char *data` as the arg this makes sense to have those casts.
> 
> Just not sure, how MB (multi-byte) strings like UTF-8 render properly … ;-)

I reviewed the `RichString_writeFromAscii()` again and it differs from `RichString_writeFromWide()` that the former is not meant to process multibyte (i.e. UTF-8) strings.

Is it okay that I add a conditional (`data[j] <= SCHAR_MAX`) to be safe? Or should this thing be an `assert()` instead? (If `RichString_writeFromAscii()` would ever need to process unsafe strings, then the answer would be the former.)

**Comment by Explorer09** [comment]
Quote the C standard text for reference.
> ### 7.4 Character handling `<ctype.h>` ###
> The header `<ctype.h>` declares several functions useful for classifying and mapping characters. In all cases the argument is an `int`, the value of which shall be representable as an `unsigned char` or shall equal the value of the macro `EOF`. If the argument has any other value, the behavior is undefined.

So our code without a cast to `unsigned char` would likely hit an undefined behavior especially when the character can be in range [-128, -2]. (`EOF` == -1)

**Comment by BenBE** [comment]
As ASCII is (against wide-spread belief) only characters `0` (NUL) through `127` (DEL), anything outside this range is technically not ASCII … Thus make it an assertion and fail hard.

**Comment by Explorer09** [comment]
@BenBE My question is, whether the function would need to handle untrusted inputs.

**Comment by BenBE** [comment]
AFAICS it's mainly used internally to speed up printing texts that we know are ASCII …

**Review by BenBE — APPROVED** [review]
LGTM.

File: RichString.c [hunk 1]
```diff
@@ -144,7 +146,8 @@ static inline int RichString_writeFromAscii(RichString* this, int attrs, const c
    int newLen = from + len;
    RichString_setLen(this, newLen);
    for (int i = from, j = 0; i < newLen; i++, j++) {
-      this->chptr[i] = (CharType) { .attr = attrs & 0xffffff, .chars = { (isprint(data[j]) ? data[j] : L'\xFFFD') } };
+      assert((unsigned char)data[j] <= SCHAR_MAX);
```
[the remark this answers is not in the document]
Comment (Explorer09) [reply RichString.c:149]: @cgzones It's not my call. (Ask @BenBE  for this.) If it's me, I would choose to be safe than sorry.

[the remark this answers is not in the document]
Comment (BenBE) [reply RichString.c:149]: @cgzones Shouldn't the command line of a process better be processed as potential MBCS; also, don't we have a non-fatal MBCS version of this for exactly this reason of writing cmdline?

[the remark this answers is not in the document]
Comment (Explorer09) [reply RichString.c:149]: @cgzones How about a fix like this? 2fa30bdb225906265937b05fad67536a35763716

Not sure if this string issue can have security impact, but the maximum impact I can imagine is an unintended crash of htop (segmentation violation internally in a `<ctype.h>` function).

[the remark this answers is not in the document]
Comment (cgzones) [reply RichString.c:149]: Yes, that's my initial concern: htop should not stop to work if any process emits some malformed (but supported by the kernel) data.

[the remark this answers is not in the document]
Comment (BenBE) [reply RichString.c:149]: Anything still missing here to go ahead with this PR?

**Review by BenBE — APPROVED** [review]
LGTM.

**Review by cgzones — APPROVED** [review]

**Pull request merged**